code/datums/storage/storage.dm

```diff
@@ -1117,10 +1055,66 @@ GLOBAL_LIST_EMPTY(cached_storage_typecaches)
 
 	modeswitch_action = null
 
+/// Updates views of all objects in storage and stretches UI to appropriate size
+/datum/storage/proc/orient_storage()
+	var/adjusted_contents = real_location.contents.len
+	var/list/datum/numbered_display/numbered_contents
+	if(numerical_stacking)
+		numbered_contents = process_numerical_display()
+		adjusted_contents = numbered_contents.len
+
+	//if the ammount of contents reaches some multiplier of the final column (and its not the last slot), let the player view an additional row
+	var/additional_row = (!(adjusted_contents % screen_max_columns) && adjusted_contents < max_slots)
+
+	var/columns = clamp(max_slots, 1, screen_max_columns)
+	var/rows = clamp(CEILING(adjusted_contents / columns, 1) + additional_row, 1, screen_max_rows)
+
+	for (var/ui_user in storage_interfaces)
+		storage_interfaces[ui_user].update_position(screen_start_x, screen_pixel_x, screen_start_y, screen_pixel_y, columns, rows)
+
+	var/current_x = screen_start_x
+	var/current_y = screen_start_y
+	var/turf/our_turf = get_turf(real_location)
+
+	if(islist(numbered_contents))
+		for(var/content_type in numbered_contents)
+			var/datum/numbered_display/numberdisplay = numbered_contents[content_type]
+
+			var/obj/item/display_sample = numberdisplay.sample_object
+			display_sample.mouse_opacity = MOUSE_OPACITY_OPAQUE
+			display_sample.screen_loc = "[current_x]:[screen_pixel_x],[current_y]:[screen_pixel_y]"
+			display_sample.maptext = MAPTEXT("<font color='white'>[(numberdisplay.number > 1)? "[numberdisplay.number]" : ""]</font>")
```

Comment:
this is basically the only difference between this for loop and the one below, could we do some concatenation?
instead of looping over EITHER the location or the numbered contents list, we should have a list variable called "place_out" or something. If we're using numbered contents we'll fill it with our display samples and set maptext, and if not we'll just set it to real_location.

would need to setup something to avoid setting maptext if we're using `numerical_stacking` but that's relatively minor in terms of complexity